[-Wunsafe-buffer-usage] Add a new warning for uses of std::span two-parameter constructors

ziqingluo-90 opened the pull request:
The PR contains two commits:
1. adding a new waring under a sub-group of `-Wunsafe-buffer-usage`
2. teach the analyzer to be quiet on some benign cases

ziqingluo-90 committed:
[-Wunsafe-buffer-usage] Add a new warning for use of two-parameter std::span constructors

Constructing `std::span` objects with the two parameter constructors
could introduce mismatched bounds information, which defeats the
purpose of using `std::span`.  Therefore, we warn every use of such
constructors.

We also plan to incrementally teach `-Wunsafe-buffer-usage` about benign
usages of those constructors.

rdar://115817781

jkorous-apple commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12075,6 +12075,9 @@ def note_unsafe_buffer_variable_fixit_together : Note<
 def note_safe_buffer_usage_suggestions_disabled : Note<
   "pass -fsafe-buffer-usage-suggestions to receive code hardening suggestions">;
+def warn_unsafe_buffer_usage_in_container : Warning<
+  "%select{the two-parameter std::span construction is unsafe as it can introduce mismatch between buffer size and the bound information}0">,
```
Looks like we should remove the select.

jkorous-apple commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -2256,6 +2256,10 @@ class UnsafeBufferUsageReporter : public UnsafeBufferUsageHandler {
+    } else if (const auto *CtorExpr = dyn_cast<CXXConstructExpr>(Operation)) {
+      S.Diag(CtorExpr->getLocation(),
+             diag::warn_unsafe_buffer_usage_in_container)
+          << 0;
```
Assuming we remove the warning message parameter we should not pass the argument here.

ziqingluo-90 committed:
[-Wunsafe-buffer-usage] Teach -Wunsafe-buffer-usage-in-container to be quiet on some benign cases

For some cases where the analyzer can tell it's safe to use
two-parameter std::span constructors, the analyzer will not report
false alarms on them.  For example,
```
std::span<T>{&var, 1} or std::span<T>{new T, 1}
```
rdar://115817781

jkorous-apple reviewed (approved):
I will just say that I didn't imagine this warning to be made part of the analysis machine we use to produce fixits and it seems like at couple places the patch struggles with the abstractions the machine uses but overall I don't think it's completely unreasonable either and we can discuss a different structure later.

ziqingluo-90 committed:
[-Wunsafe-buffer-usage][NFC] fix clang format
